crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -2564,25 +2558,143 @@ mod test_finalize_block {
+                Fee {
+                    amount_per_gas_unit: DenominatedAmount::native(1.into()),
+                    token: shell.state.in_mem().native_token.clone(),
+                },
+                keypair.ref_to(),
+                Epoch(0),
+                WRAPPER_GAS_LIMIT.into(),
+                None,
+            ))));
+        wrapper.header.chain_id = shell.chain_id.clone();
+        wrapper.set_code(Code::new(tx_code, None));
+        let key = Key::from(Address::from(&keypair_2.ref_to()).to_db_key())
+            .push(&"test".to_string())
+            .unwrap();
+        wrapper.set_data(Data::new(
+            TxWriteData {
+                key,
+                value: "test".as_bytes().to_vec(),
+            }
+            .serialize_to_vec(),
+        ));
+
+        let mut new_wrapper = wrapper.clone();
+        new_wrapper.update_header(TxType::Wrapper(Box::new(WrapperTx::new(
+            Fee {
+                amount_per_gas_unit: DenominatedAmount::native(1.into()),
+                token: shell.state.in_mem().native_token.clone(),
+            },
+            keypair_2.ref_to(),
+            Epoch(0),
+            WRAPPER_GAS_LIMIT.into(),
+            None,
+        ))));
+        new_wrapper.add_section(Section::Authorization(Authorization::new(
+            vec![new_wrapper.raw_header_hash()],
+            [(0, keypair_2.clone())].into_iter().collect(),
+            None,
+        )));
+        // This is a signature coming from the wrong signer which will be
+        // rejected by the vp
+        wrapper.add_section(Section::Authorization(Authorization::new(
+            vec![wrapper.raw_header_hash()],
+            [(0, keypair.clone())].into_iter().collect(),
+            None,
+        )));
+        new_wrapper.add_section(Section::Authorization(Authorization::new(
+            new_wrapper.sechashes(),
+            [(0, keypair_2)].into_iter().collect(),
+            None,
+        )));
+        wrapper.add_section(Section::Authorization(Authorization::new(
+            wrapper.sechashes(),
+            [(0, keypair)].into_iter().collect(),
+            None,
+        )));
+
+        let mut processed_txs: Vec<ProcessedTx> = vec![];
+        for tx in [&wrapper, &new_wrapper] {
+            processed_txs.push(ProcessedTx {
+                tx: tx.to_bytes().into(),
+                result: TxResult {
+                    code: ResultCode::Ok.into(),
+                    info: "".into(),
+                },
+            })
+        }
+
+        // merkle tree root before finalize_block
+        let root_pre = shell.shell.state.in_mem().block.tree.root();
+
+        let event = &shell
+            .finalize_block(FinalizeBlock {
+                txs: processed_txs,
+                ..Default::default()
+            })
+            .expect("Test failed");
+
+        // the merkle tree root should not change after finalize_block
+        let root_post = shell.shell.state.in_mem().block.tree.root();
```

Comment:
This is probably me forgetting details, but since the second tx should pass, wouldn't its hash be added to replay prot and change the merkle tree?

Reply:
The reprot entries of the db are not merklized

Reply:
ah right.